cinder/trace_utils.py

```diff
@@ -0,0 +1,194 @@
+
+    @staticmethod
+    def trace_with_file_info(context, action, resource, message,
+                             filename, line_number, function, failed=False):
+        # Do not call this directly.
+        # Call trace_utils.trace_with_file_info() instead.
+
+        # We only want the frame of the caller
+        # we should always be called from the trace() method in this module
+        # not called directly in this static method
+        LogTrace._trace_with_info(
+            context, action, resource, message,
+            filename, line_number, function, failed
+        )
+
+
+def trace(context, action, resource, message, failed=False):
+    """For now we only implement LogTrace.
+
+    TODO(waboring): add rabbitmq trace? to send entries to a msg queue
+    or add: DBtrace to send traces to the DB instead?
+    """
+    LogTrace.trace(context, action, resource, message, failed)
+
+
+def trace_with_info(context, action, resource, message, file, line_number,
+                    function, failed=False):
+    """For now we only implement LogTrace.
+
+    TODO(waboring): add rabbitmq trace? to send entries to a msg queue
+    or add: DBtrace to send traces to the DB instead?
+    """
+    LogTrace.trace_with_file_info(context, action, resource, message, file,
+                                  line_number, function, failed)
+
+
+def trace_decorator(action):
+    """Decorator to automatically handle exceptions raised as failures.
+
+       Place this decorator on a function that you want to mark as an
+       action failure and the trace_utils tracing will get called
+       for the action.
+
+       @trace_decorator(trace_utils.ACTION_VOLUME_ATTACH)
+       def initialize_connection(....)
+
+       If initialize_connection raises an exception then you will get a
+       trace_utils.trace called with action of ACTION_VOLUME_ATTACH
+       and a failure.
+
+    """
+    @decorator.decorator
+    def inner(func, *args, **kwargs):
+        # Find the context and the volume/backup object
+        context = None
+        resource = None
+        valid_resource_names = [
+            'volume', 'backup', 'snapshot'
+        ]
+        valid_context_names = [
+            'context', 'ctxt'
+        ]
+        resource = None
+        context = None
+        call_args = inspect.getcallargs(func, *args, **kwargs)
+        for key in call_args:
+            if key in valid_resource_names:
+                resource = call_args[key]
+            elif (key in valid_context_names and
+                    isinstance(call_args[key], cinder_context.RequestContext)):
+                context = call_args[key]
+
+        trace(context, action, resource, "called")
+
+        try:
+            return func(*args, **kwargs)
+        except Exception:
+            with excutils.save_and_reraise_exception() as exc:
+                tl = traceback.extract_tb(exc.tb)
+                i = tl[1]
+                message = str(exc.value)
```

Comment:
```
$ python3 /tmp/bla.py 
['__cause__', '__class__', '__context__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '__suppress_context__', '__traceback__', 'args', 'with_traceback']
Traceback (most recent call last):
  File "/tmp/bla.py", line 2, in <module>
    raise RuntimeError('asdf')
RuntimeError: asdf

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/bla.py", line 5, in <module>
    print(e.value)
AttributeError: 'RuntimeError' object has no attribute 'value'
```
```
try:
    raise RuntimeError('asdf')
except Exception as e:
    print(dir(e))
    print(e.value)
```

Exceptions don't have a `value` attribute. You're getting it, because you're not accessing an exception, but the [`save_and_reraise_exception`](https://github.com/openstack/oslo.utils/blob/1b8bafb391b82665b99b3da8205402ec2cd495d1/oslo_utils/excutils.py#L146) instance. `value` equals the second item from [`sys.exc_info()`](https://docs.python.org/3/library/sys.html#sys.exc_info), which is the original exception getting raised. So basically, you're doing `str(e)` on the original exception. Everything is fine then :)